rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+        Note over Local Machine: user taps YubiKey<br/>to sign MFA challenge
+        Local Machine->>Teleport Auth: rpc UpdateHeadlessAuthRequestState<br/>with signed MFA challenge response
+    end
+
+    Teleport Auth->>-Teleport Proxy: unblock on request state change
+    Teleport Proxy->>Teleport Auth: rpc GenerateUserCerts<br/>(MFA-verified, 1 minute TTL)
+    Teleport Auth->>Teleport Proxy: 
+    Teleport Proxy->>Remote Machine: user certificates
+    Note over Remote Machine: Connect to user@node01
+```
+
+This flow can be broken down into three parts: headless login initiation, local authentication, and certificate retrieval.
+
+#### Headless login initiation
+
+First, the client initiates headless login through the web proxy endpoint `POST /webapi/login/headless`. The client provides a request UUID, the command being requested (e.g. `tsh ssh user@host`), and normal login parameters (client public key, proxy address, etc.). The Proxy generates a request token which will be used to authenticate the proxy in the final steps of headles authentication.
+
+The request details and request token are saved on the Auth server in a `HeadlessAuthRequest` object, with the backend prefix `/headless_auth_requests/`. It will be saved with a 1 minute TTL, by which point the user should have completed the headless authentication flow. The request will begin in the pending state, to be later approved/denied by the user locally.
+
+```go
+type HeadlessAuthRequest struct {
+    ID string
+    // User is a teleport username
+    User string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    // Command is the client command being requested
+    Command string
+    // State is the request state (pending, approved, or denied)
+    State string
+}
+```
+
+#### Local authentication
+
+In parallel to headless login initiation, the client will generate a Teleport web Proxy URL for the client to complete headless authentication: `https://proxy.example.com/headless/<request_id>/approve`. The URL is shared with the user so they can locally authenticate the auth request from their local browser.
+
+When the user runs the command or opens the URL locally, their local login session will be used to connect to the Teleport Auth server. If the user is not yet logged in, they will be prompted to login with MFA as usual.
+
+Once connected, the user can view the request details and either approve or deny the request:
+
+* If the user approves the request, they will need to pass an MFA challenge to update the request to the approved state.
+* If the user denies the request, the request will be updated to the denied state.
+
+#### Certificate retrieval
+
+After the headless login is initiated, the request will wait until local authentication is complete. This will be handled by the Teleport Proxy by using a resource watcher to wait until the `HeadlessAuthRequest` object is updated to the approved or denied state, or it expires.
```

Comment:
I would double-check with the scalability team, just to be sure.

Reply:
Would that be @rosstimothy? Can you give this RFD a brief look when you get a chance? Specifically, are the following approaches scalable:
 - Adding 1 minute TTL to `HeadlessAuthentication` resource
   - I know we've had issues with too many expiration trackers for Session Trackers in the past.
 - Watching for `HeadlessAuthentication` updates from Proxy -> Auth

The primary concern here is that `/webapi/login/headless` is unauthenticated and subject to DoS attacks, though it will be rate limited. Both the creation and watch of `HeadlessAuthentication` occur before the end user has been authenticated.